File: tests/ClaudeStatusBar.Tests/GoldenReplayTests.cs
using ClaudeStatusBar.Data;
using ClaudeStatusBar.Model;
using Xunit;

namespace ClaudeStatusBar.Tests;

/// <summary>
/// Replays the real captured log (tests/fixtures/window-shape-2026-09-11.csv,
/// a verbatim copy of the machine's actual logs\window-shape.csv -- see
/// docs/forecast-and-states.md's evidence section) through the exact ingest
/// path a live poll uses. That file contains real replica regressions
/// (15->14, 19->18, both with a fresh and with a previously-seen fingerprint),
/// a real one-second resets_at jitter (2026-09-09, decision 1's motivating
/// case), and a stale, unrelated older window (also 2026-09-09, a different
/// boot/session) that a genuine rollover must still separate from the current
/// one.
/// </summary>
public class GoldenReplayTests
{
    static string FixturePath => Path.Combine(AppContext.BaseDirectory, "fixtures", "window-shape-2026-09-11.csv");

    [Fact]
    public void RealFixture_ReplaysCleanly_NoRollover_RegressionsIgnored_PositiveRate_SensibleState()
    {
        IReadOnlyList<CsvReplay.RawRow> allRows = CsvReplay.ReadRows(FixturePath);
        Assert.True(allRows.Count > 50, $"expected the golden fixture to have many rows, got {allRows.Count}");

        // "Now" = the time of the last poll in the file; its resets_at defines the current window key.
        CsvReplay.RawRow lastRow = allRows[^1];
        DateTimeOffset now = lastRow.UtcIso;
        Assert.NotNull(lastRow.SessionResetsAtRaw);
        QuotaTimeUtil.TryParseResetsAt(lastRow.SessionResetsAtRaw, out DateTimeOffset sessionResetsAt);
        DateTimeOffset sessionKey = QuotaTimeUtil.TruncateToSecond(sessionResetsAt);

        IReadOnlyList<CsvReplay.MatchedRow> matched = CsvReplay.FilterForWindow(
            allRows, WindowKind.Session, sessionKey, QuotaWindows.SessionMinutes, now);

        // The fixture also carries an older, unrelated window (a different boot):
        // the filter must have actually excluded those rows, not passed everything through.
        Assert.True(matched.Count < allRows.Count);
        Assert.True(matched.Count > 50);
        Assert.All(matched, row => Assert.True(row.UtcIso.Year == 2026 && row.UtcIso.Month == 9 && row.UtcIso.Day == 11));

        var tracker = new WindowTracker(QuotaWindows.SessionMinutes, halfLifeMinutes: 15);
        double previousEnvelope = 0.0;
        foreach (CsvReplay.MatchedRow row in matched)
        {
            tracker.Ingest(row.Pct, row.ResetsAtRaw, row.UtcIso, row.MonoMs); // some rows are legitimate cached duplicates (rejected); must never throw

            // Test critique: assert the monotone-envelope invariant PER OBSERVATION, not just
            // once at the end -- a broad final-value check alone cannot catch an intermediate
            // dip that a later, higher sample happens to paper over.
            Assert.True(tracker.EnvelopeP >= previousEnvelope, $"envelope dipped from {previousEnvelope} to {tracker.EnvelopeP} at {row.UtcIso:O}");
            previousEnvelope = tracker.EnvelopeP;
        }

        // No rollover was ever detected during replay (by construction, every matched row
        // already shares the one target window key). WindowKey now tracks the freshest raw
        // deadline (decision 1), not a truncated-to-the-second value, so compare truncated.
        Assert.Null(tracker.RolloverAt);
        Assert.NotNull(tracker.WindowKey);
        Assert.Equal(sessionKey, QuotaTimeUtil.TruncateToSecond(tracker.WindowKey!.Value));

        // Regressions ignored: the envelope equals the true max, not the last value. Review
        // critique: "a percentage below the final maximum does not establish a regression;
        // an entirely increasing sequence passes." Find an actual adjacent, non-increasing
        // pair in the SOURCE data (a genuine regression), and confirm the envelope did not
        // dip there.
        double expectedMax = matched.Max(r => r.Pct);
        Assert.Equal(expectedMax, tracker.EnvelopeP);

        bool foundGenuineRegression = false;
        for (int i = 1; i < matched.Count; i++)
        {
            if (matched[i].Pct < matched[i - 1].Pct)
            {
                foundGenuineRegression = true;
                break;
            }
        }
        Assert.True(foundGenuineRegression, "expected the known real replica regressions (e.g. 15->14, 19->18) to be present in the source data");

        WindowSnapshot snap = tracker.ComputeSnapshot(now, lastRow.MonoMs, isWeekly: false);
        Assert.True(snap.Forecast.HasValue);
        ForecastResult f = snap.Forecast!.Value;

        // Stronger than "positive and Safe-or-Tight" (review critique of the old assertion,
        // which is a weak plausibility check, not an independently justified oracle): with
        // ~20% used and hours left to reset, the model's own algebra requires depletion to
        // land AFTER the reset, i.e. zero shortfall -- which alone rules out DryEarly (its
        // only other trigger, rem <= 3, plainly does not hold at ~20% used either).
        Assert.True(f.RatePctPerMin is > 0.0 and < 5.0, $"expected a small positive burn rate for this fixture, got {f.RatePctPerMin}");
        Assert.Equal(0.0, f.ShortfallMinutes);
        Assert.True(f.RemainingPct > 50.0);

        QuotaState state = RawStateClassifier.Classify(tracker.EnvelopeP, f, QuotaWindows.SessionMinutes, snap.Refused);
        Assert.True(state is QuotaState.Safe or QuotaState.Tight, $"expected a plausible verdict for ~{tracker.EnvelopeP}% used with hours to reset, got {state}");
    }

    /// <summary>
    /// Decision 1's exact motivating case, from the real fixture: resets_at moves
    /// 19:10:00.218248 -&gt; 19:09:59.450402 (~0.77s EARLIER) between the first two
    /// 2026-09-09 polls, while usage rises 85% -&gt; 94%. Under the old truncate-to-
    /// the-second key comparison this looked like a regressed window key and the
    /// 94% sample was silently dropped, permanently retaining 85% while the server
    /// had already reported 94%.
    /// </summary>
    [Fact]
    public void Fixture_2026_09_09_JitterRows_KeepsTheHigherSample_NotDropped()
    {
        IReadOnlyList<CsvReplay.RawRow> allRows = CsvReplay.ReadRows(FixturePath);
        List<CsvReplay.RawRow> sept9Rows = allRows
            .Where(r => r.UtcIso.Year == 2026 && r.UtcIso.Month == 9 && r.UtcIso.Day == 9)
            .OrderBy(r => r.UtcIso)
            .ToList();
        Assert.True(sept9Rows.Count >= 2, "expected the fixture to carry the 2026-09-09 jitter rows");

        var tracker = new WindowTracker(QuotaWindows.SessionMinutes, halfLifeMinutes: 15);
        foreach (CsvReplay.RawRow row in sept9Rows)
        {
            Assert.NotNull(row.SessionResetsAtRaw);
            Assert.NotNull(row.SessionPct);
            tracker.Ingest(row.SessionPct!.Value, row.SessionResetsAtRaw, row.UtcIso, row.MonoMs); // must never throw
        }

        Assert.Equal(94.0, tracker.EnvelopeP); // kept, not dropped
        Assert.Null(tracker.RolloverAt); // the ~0.77s jitter must never have been mistaken for a rollover
    }

    /// <summary>
    /// Review critique: the main golden-replay test "cannot discover reset jitter" because
    /// it pre-filters to one exact window key before replay -- exactly hiding the case the
    /// fixture demonstrates. This replays the ENTIRE fixture, unfiltered, through one
    /// tracker in chronological order, and lets the tracker's own decision-1 logic (not
    /// external key-matching) tell same-window jitter (2026-09-09) apart from a genuine,
    /// much-later window (2026-09-11, a different boot two days on).
    /// </summary>
    [Fact]
    public void FullFixture_ReplayedContinuously_WithoutPreFiltering_SeparatesJitterFromTheGenuineRollover()
    {
        IReadOnlyList<CsvReplay.RawRow> allRows = CsvReplay.ReadRows(FixturePath);
        List<CsvReplay.RawRow> chronological = allRows.OrderBy(r => r.UtcIso).ToList();

        var tracker = new WindowTracker(QuotaWindows.SessionMinutes, halfLifeMinutes: 15);
        int rolloverEvents = 0;
        DateTimeOffset? lastSeenRolloverAt = null;

        foreach (CsvReplay.RawRow row in chronological)
        {
            if (row.SessionPct is null || row.SessionResetsAtRaw is null) continue;
            tracker.Ingest(row.SessionPct.Value, row.SessionResetsAtRaw, row.UtcIso, row.MonoMs); // must never throw
            if (tracker.RolloverAt != lastSeenRolloverAt)
            {
                rolloverEvents++;
                lastSeenRolloverAt = tracker.RolloverAt;
            }
        }

        // Exactly one genuine rollover (2026-09-09's window -> 2026-09-11's): the within-
        // 2026-09-09 jitter must not have counted as a second one.
        Assert.Equal(1, rolloverEvents);
        Assert.True(tracker.EnvelopeP > 0);
    }
}

File: src/Mac/Tests/ClaudeQuotaCoreTests/GoldenReplayTests.swift
import Foundation
import Testing

@testable import ClaudeQuotaCore

/// The Swift half of M2's acceptance criterion: replay `tests/fixtures/window-shape-2026-09-11.csv`
/// — the *same file* the C# `GoldenReplayTests` reads — through the exact ingest path a live poll
/// uses, and reach the same verdicts.
///
/// That file contains real replica regressions (15→14, 19→18, both with a fresh and with a
/// previously-seen fingerprint), a real one-second `resets_at` jitter (2026-09-09, decision 1's
/// motivating case), and a stale, unrelated older window (a different boot) that a genuine
/// rollover must still separate from the current one.
@Suite("Golden replay")
struct GoldenReplayTests {

    @Test("the real fixture replays cleanly: no rollover, regressions ignored, a sensible verdict")
    func realFixtureReplaysCleanly() throws {
        let allRows = Fixtures.rows(Fixtures.windowShape)
        #expect(allRows.count > 50, "expected the golden fixture to have many rows, got \(allRows.count)")

        // "Now" = the time of the last poll in the file; its resets_at defines the current key.
        let lastRow = try #require(allRows.last)
        let now = lastRow.utcIso
        let sessionResetsAt = try #require(QuotaTimeUtil.parseResetsAt(lastRow.sessionResetsAtRaw))
        let sessionKey = QuotaTimeUtil.truncateToSecond(sessionResetsAt)

        let matched = CsvReplay.filterForWindow(allRows, kind: .session, windowKey: sessionKey,
                                                windowMinutes: QuotaWindows.sessionMinutes, now: now)

        // The fixture also carries an older, unrelated window (a different boot): the filter must
        // have actually excluded those rows, not passed everything through.
        #expect(matched.count < allRows.count)
        #expect(matched.count > 50)
        for row in matched {
            let ymd = row.utcIso.utcYMD
            #expect(ymd.year == 2026 && ymd.month == 9 && ymd.day == 11)
        }

        let tracker = WindowTracker(windowMinutes: QuotaWindows.sessionMinutes, halfLifeMinutes: 15)
        var previousEnvelope = 0.0
        for row in matched {
            // Some rows are legitimate cached duplicates (rejected); must never throw.
            tracker.ingest(pct: row.pct, resetsAtRaw: row.resetsAtRaw, utcNow: row.utcIso, monoMs: row.monoMs)

            // The monotone-envelope invariant PER OBSERVATION, not just once at the end: a broad
            // final-value check alone cannot catch an intermediate dip that a later, higher
            // sample happens to paper over.
            #expect(tracker.envelopeP >= previousEnvelope,
                    "envelope dipped from \(previousEnvelope) to \(tracker.envelopeP)")
            previousEnvelope = tracker.envelopeP
        }

        // No rollover during replay (by construction, every matched row shares one key).
        #expect(tracker.rolloverAt == nil)
        let windowKey = try #require(tracker.windowKey)
        #expect(QuotaTimeUtil.truncateToSecond(windowKey) == sessionKey)

        // Regressions ignored: the envelope equals the true max, not the last value.
        let expectedMax = try #require(matched.map(\.pct).max())
        #expect(tracker.envelopeP == expectedMax)

        // A percentage below the final maximum does not by itself establish a regression, so
        // confirm an actual adjacent non-increasing pair exists in the SOURCE data.
        var foundGenuineRegression = false
        for index in 1..<matched.count where matched[index].pct < matched[index - 1].pct {
            foundGenuineRegression = true
            break
        }
        #expect(foundGenuineRegression,
                "expected the known real replica regressions (15→14, 19→18) in the source data")

        let snap = tracker.computeSnapshot(utcNow: now, monoMs: lastRow.monoMs, isWeekly: false)
        let forecast = try #require(snap.forecast)

        // Stronger than "positive and Safe-or-Tight": with ~20 % used and hours to reset the
        // model's own algebra requires depletion to land AFTER the reset, i.e. zero shortfall —
        // which alone rules out DryEarly (its only other trigger, rem <= 3, plainly does not
        // hold at ~20 % either).
        #expect(forecast.ratePctPerMin > 0.0 && forecast.ratePctPerMin < 5.0,
                "expected a small positive burn rate, got \(forecast.ratePctPerMin)")
        #expect(forecast.shortfallMinutes == 0.0)
        #expect(forecast.remainingPct > 50.0)

        let state = RawStateClassifier.classify(usedPct: tracker.envelopeP, forecast: forecast,
                                                windowMinutes: QuotaWindows.sessionMinutes,
                                                refused: snap.refused)
        #expect(state == .safe || state == .tight,
                "expected a plausible verdict for ~\(tracker.envelopeP) % used with hours to reset, got \(state)")
    }

    /// Decision 1's exact motivating case, from the real fixture: `resets_at` moves
    /// `19:10:00.218248` → `19:09:59.450402` (~0.77 s EARLIER) between the first two 2026-09-09
    /// polls, while usage rises 85 % → 94 %. Under the old truncate-to-the-second key comparison
    /// this looked like a regressed window key and the 94 % sample was silently dropped,
    /// permanently retaining 85 % while the server had already reported 94 %.
    @Test("the 2026-09-09 jitter rows keep the higher sample instead of dropping it")
    func jitterRowsKeepTheHigherSample() throws {
        let allRows = Fixtures.rows(Fixtures.windowShape)
        let sept9 = allRows
            .filter { let d = $0.utcIso.utcYMD; return d.year == 2026 && d.month == 9 && d.day == 9 }
            .sorted { $0.utcIso < $1.utcIso }
        #expect(sept9.count >= 2, "expected the fixture to carry the 2026-09-09 jitter rows")

        let tracker = WindowTracker(windowMinutes: QuotaWindows.sessionMinutes, halfLifeMinutes: 15)
        for row in sept9 {
            let pct = try #require(row.sessionPct)
            #expect(row.sessionResetsAtRaw != nil)
            tracker.ingest(pct: pct, resetsAtRaw: row.sessionResetsAtRaw, utcNow: row.utcIso, monoMs: row.monoMs)
        }

        #expect(tracker.envelopeP == 94.0, "the higher sample was dropped")
        #expect(tracker.rolloverAt == nil, "the ~0.77 s jitter was mistaken for a rollover")
    }

    /// The main replay pre-filters to one window key, which hides the very case the fixture
    /// demonstrates. This replays the ENTIRE fixture, unfiltered, through one tracker in
    /// chronological order, and lets the tracker's own decision-1 logic — not external key
    /// matching — tell same-window jitter (2026-09-09) apart from a genuine, much later window
    /// (2026-09-11, a different boot two days on).
    @Test("the full fixture separates jitter from the one genuine rollover")
    func fullFixtureSeparatesJitterFromRollover() {
        let chronological = Fixtures.rows(Fixtures.windowShape).sorted { $0.utcIso < $1.utcIso }

        let tracker = WindowTracker(windowMinutes: QuotaWindows.sessionMinutes, halfLifeMinutes: 15)
        var rolloverEvents = 0
        var lastSeenRolloverAt: Date?

        for row in chronological {
            guard let pct = row.sessionPct, row.sessionResetsAtRaw != nil else { continue }
            tracker.ingest(pct: pct, resetsAtRaw: row.sessionResetsAtRaw, utcNow: row.utcIso, monoMs: row.monoMs)
            if tracker.rolloverAt != lastSeenRolloverAt {
                rolloverEvents += 1
                lastSeenRolloverAt = tracker.rolloverAt
            }
        }

        // Exactly one genuine rollover (2026-09-09's window → 2026-09-11's): the within-09-09
        // jitter must not have counted as a second one.
        #expect(rolloverEvents == 1, "expected exactly one rollover, saw \(rolloverEvents)")
        #expect(tracker.envelopeP > 0)
    }

    /// Cross-implementation sanity on the shared file itself: if the CSV reader disagrees with
    /// the C# one about how many rows are well-formed, every verdict downstream is comparing
    /// different data and the whole acceptance criterion is void.
    @Test("the shared fixture parses to the row counts the C# suite sees")
    func fixtureParsesIdentically() {
        let rows = Fixtures.rows(Fixtures.windowShape)
        // window-shape-2026-09-11.csv is 128 lines: 1 header + 127 data rows, all well-formed.
        #expect(rows.count == 127, "expected 127 parsed rows, got \(rows.count)")

        let rollover = Fixtures.rows(Fixtures.rollover)
        // rollover-2026-09-11.csv is 37 lines: 1 header + 36 data rows.
        #expect(rollover.count == 36, "expected 36 parsed rows, got \(rollover.count)")
    }
}
